### PR TITLE
feat: [Magic Link Login] add placeholder to `invalidEmail` to return user email.

### PR DESCRIPTION
**Pull request by warcooft**
**Description**
This pull request improves existing translations by returning the user's email if the corresponding email is not found.  This update aims to make communication more engaging and user-friendly. By returning the email value in the error message, it will help users to confirm and correct the email sent because in general users often make typos.

**Preview**

**Screenshot 2024-07-17 at 14 41 47**

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

**Comment by kenjis** [comment]
Please fix errors in GitHub Action checks, and add test code.
Reference: https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/pull_request.md

**Comment by warcooft** [comment]
@kenjis all language files need to be updated but I don't understand other languages. Should I use english for other files as default language to be translated later by respective authors?

**Comment by kenjis** [comment]
@warcooft 
See https://github.com/codeigniter4/shield/commit/2dc12527acc30f3ad10ddc5e41173679a78ecbb5#diff-2b85118f8a629f90163570fda562d9379b8ee25dafc16e8748e7eb892884ddd2

**Comment by warcooft** [comment]
Ahh, OK. I will update later.

File: src/Controllers/MagicLinkController.php [hunk 1]
```diff
@@ -92,7 +92,7 @@ public function loginAction()
         $user  = $this->provider->findByCredentials(['email' => $email]);
 
         if ($user === null) {
-            return redirect()->route('magic-link')->with('error', lang('Auth.invalidEmail'));
+            return redirect()->route('magic-link')->with('error', lang('Auth.invalidEmail', [$email]));
```
Comment (kenjis) [inline src/Controllers/MagicLinkController.php:95]: The session variable `error` is a HTML string, because it is not escaped.
So there is a room to XSS attacks with email like `"><svg/onload=confirm(1)>"@gmail.com`.
https://github.com/codeigniter4/shield/blob/a38b3ccc5d70862ffc27b470a76cb4156f4d1a70/src/Views/magic_link_form.php#L13

Comment (kenjis) [reply src/Controllers/MagicLinkController.php:95]: Is it better that we escape all session `error` `errors` in views?

Comment (warcooft) [reply src/Controllers/MagicLinkController.php:95]: agree! it should be escaped.

Comment (warcooft) [reply src/Controllers/MagicLinkController.php:95]: should I update it too or do I need to make a new PRs?

Comment (kenjis) [reply src/Controllers/MagicLinkController.php:95]: Escape in the magic link view in this PR, because this PR makes output with email address.

Please send another PR for other views.

Comment (warcooft) [reply src/Controllers/MagicLinkController.php:95]: ahh, okay.

File: src/Language/en/Auth.php [hunk 2]
```diff
@@ -25,7 +25,7 @@
     'badToken'              => 'The access token is invalid.',
     'oldToken'              => 'The access token has expired.',
     'noUserEntity'          => 'User Entity must be provided for password validation.',
-    'invalidEmail'          => 'Unable to verify the email address matches the email on record.',
+    'invalidEmail'          => 'Unable to verify the email address matches the email {0}.',
```
Comment (kenjis) [inline src/Language/en/Auth.php]: Is this better?
```suggestion
    'invalidEmail'          => 'Unable to verify the email address matches the email "{0}".',
```

Screenshot 2024-07-19 11 26 08

**Comment by kenjis** [comment]
Escape this:
https://github.com/codeigniter4/shield/blob/18294434912d7b91aead51b5d6c425d9d5321692/src/Views/magic_link_form.php#L22

File: src/Language/en/Auth.php (hunk 2 above)
Comment (warcooft) [reply src/Language/en/Auth.php]: ahh, that's much better. I didn't even think about it.

**Comment by warcooft** [comment]
I'm in a meeting now, I'll update again later.

**Review by kenjis — CHANGES_REQUESTED** [review]
Escape this, too.
https://github.com/codeigniter4/shield/blob/18294434912d7b91aead51b5d6c425d9d5321692/src/Views/magic_link_form.php#L22

**Comment by kenjis** [comment]
@warcooft We don't use merge commits in PR branches. See GitHub Actions.
Can you remove it?
How to update your PR branch: https://github.com/codeigniter4/CodeIgniter4/blob/develop/contributing/workflow.md#updating-your-branch

**Comment by warcooft** [comment]
ups, my bad, okay I'll rebase.

**Comment by warcooft** [comment]
It seems to be fine now, please check it.

File: composer.json [hunk 3]
```diff
@@ -39,7 +39,7 @@
         "phpstan/extension-installer": "^1.3",
         "phpstan/phpstan-strict-rules": "^1.5",
         "phpunit/phpunit": "^9.6",
-        "rector/rector": "1.2.0"
+        "rector/rector": "1.2.1"
```
Comment (kenjis) [inline composer.json]: Remove this unrelated change.

Comment (kenjis) [reply composer.json]: Rector is already v1.2.1 on `develop`.
https://github.com/codeigniter4/shield/blob/develop/composer.json#L42

```
git fetch upstream
git rebase upstream/develop
```

Comment (warcooft) [reply composer.json]: done

**Comment by warcooft** [comment]
Where can I find that key in the test case?

```
19) Tests\Language\UkrainianTranslationTest::testAllIncludedLanguageKeysAreTranslated with data set "uk" ('uk')
Failed asserting that the translated language key "Auth.invalidEmail" in "uk" locale differs from the original keys in the main repository.
Failed asserting that an array is empty.
```

**Comment by kenjis** [comment]
> Where can I find that key in the test case?

See https://github.com/codeigniter4/shield/pull/1145/files#diff-12beddb17619112c65bd21202ce1fe8bcd291beb594aa101c90f09b174289fc7R28

> Failed asserting that the translated language key "Auth.invalidEmail" in "uk" locale differs from the original keys in the main repository.

It is difficult to understand, but it tells "Auth.invalidEmail" in "uk" must be different from `en`.

**Comment by warcooft** [comment]
I see, I forgot to add something.

**Review by warcooft — COMMENTED** [review]
done

**Review by kenjis — APPROVED** [review]
LGTM!

**Review by datamweb — APPROVED** [review]
@warcooft thank you!

Sorry for the delay.
